### PR TITLE
Oleh multinomial

### PR DESCRIPTION
**Pull request by olegsemeniv**
multinomial (categorical) random number generator both cpu and cuda

### CODE REVIEW

**Review by shyrma — DISMISSED** [review]

File: libnd4j/include/ops/declarable/helpers/cuda/random.cu [hunk 1]
```diff
@@ -248,6 +250,113 @@ namespace helpers {
...
+     const int blocksPerGrid = (batchValue * numOfSamples + threadsPerBlock - 1) / threadsPerBlock;
+    
+     PointersManager manager(context, "fillMultinomial");
+     graph::RandomGenerator *devRng;
+     auto err = cudaMalloc(&devRng, sizeof(graph::RandomGenerator));
+     if (err != 0) {
+         cuda_exception::build("fillRandomMultiNomial: Cannot allocate device memory for random generator due error", err);
+     }
+
+     err = cudaMemcpy(devRng, &rng, sizeof(graph::RandomGenerator), cudaMemcpyHostToDevice);
```
Comment (raver119) [inline libnd4j/include/ops/declarable/helpers/cuda/random.cu]: Please use cudaMemcpyAsync with the same stream as used for kernel launch

File: libnd4j/include/ops/declarable/generic/random/multinomial.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+            auto dimA = (0 == dimC) ? 1 : 0;
+            if (1 == input->sizeAt(dimA)) {
+                *output = 0;
+                return Status::OK();
+            }
+
+            auto rng = block.randomGenerator();
+            if (argSize > 3) {
+                rng.setStates(static_cast<Nd4jLong>(INT_ARG(3)), static_cast<Nd4jLong>(INT_ARG(3)));
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/random/multinomial.cpp]: As discussed before - don't do this please. Just use RNG as is, and assume it's propagated properly.

**Review by raver119 — CHANGES_REQUESTED** [review]
Few changes required

File: libnd4j/include/ops/declarable/generic/random/multinomial.cpp (hunk 2 above)
Comment (raver119) [reply libnd4j/include/ops/declarable/generic/random/multinomial.cpp]: P.s. for tests purposes use Context directly, and set both rng states there.

File: libnd4j/include/ops/declarable/helpers/cuda/random.cu (hunk 1 above)
Comment (olegsemeniv) [reply libnd4j/include/ops/declarable/helpers/cuda/random.cu]: done

File: libnd4j/include/ops/declarable/generic/random/multinomial.cpp (hunk 2 above)
Comment (olegsemeniv) [reply libnd4j/include/ops/declarable/generic/random/multinomial.cpp]: done

File: libnd4j/include/ops/declarable/generic/random/multinomial.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+
+        DECLARE_SHAPE_FN(random_multinomial) {
+
+            const int argSize = block.getIArguments()->size();
+            REQUIRE_TRUE(argSize >= 1, 0, "RANDOM_MULTINOMIAL OP: Have to be specified at least one sample,"
+                                          " but got no argumets instead.");
+
+            auto nSamples = INT_ARG(0);
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/random/multinomial.cpp]: According to what i see in op spec, nSamples in TF is dynamic, so it should be mapped as another INPUT_VARIABLE

**Review by raver119 — CHANGES_REQUESTED** [review]

File: libnd4j/include/ops/declarable/generic/random/multinomial.cpp (hunk 3 above)
Comment (raver119) [reply libnd4j/include/ops/declarable/generic/random/multinomial.cpp]: https://github.com/KonduitAI/deeplearning4j/blob/b5f0ec072f3fd0da566e32f82c0e43ca36553f39/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/resources/ops.proto#L18430-L18433

Comment (olegsemeniv) [reply libnd4j/include/ops/declarable/generic/random/multinomial.cpp]: done

**Review by raver119 — APPROVED** [review]
LGTM 👍

**Pull request merged**